### PR TITLE
[IMAGING-351] ExifRewriterRoundtripTest fixed, completed & enabled

### PR DESCRIPTION
**Pull request by StefanOltmann**
This completes the work started in PR #275.

This unit test will prove that the fix in PR #359 / #203 is correct.

Drop this [testfile.jpg](https://github.com/apache/commons-imaging/assets/9299303/2826b4ee-473a-45fb-b9fc-2168b9ae3fca) into `src/test/resources/data/images/jpg` to make the test fail.

### CODE REVIEW

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (StefanOltmann) [inline src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:71]: The original test did not compare the bytes. This is needed to really detect data truncation.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:162]: This was the reason for the orientation value change that @kinow mentioned in https://github.com/apache/commons-imaging/pull/275#pullrequestreview-1305362377

The byte order must be preserved or else the interpretation of bytes will change.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:134]: This `toString()` comparison test is useless.

The TiffWriter does change the order of fields in an attempt to optimize for space. This results in fields being reordered and unnecessary offset fields to be dropped.

You only can compare the `TiffOutputSet`.

I assume the author did this to compare values. This is why I introduced the value comparison API to TiffOutputField.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:173]: This toString() comparison test is useless.

The TiffWriter does change the order of fields in an attempt to optimize for space. This results in fields being reordered and unnecessary offset fields to be dropped.

You only can compare the TiffOutputSet.

I assume the author did this to compare values. This is why I introduced the value comparison API to TiffOutputField.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:163]: This was the reason for the orientation value change that @kinow mentioned in https://github.com/apache/commons-imaging/pull/275#pullrequestreview-1305362377

The byte order must be preserved or else the interpretation of bytes will change.

**Comment by garydgregory** [comment]
Hello @StefanOltmann 

I don't think I understand what you are saying here. What I see in this PR is:
- the addition of a utility method to main and 
- changes to a test case
- the claim that this PR proves that another PR (#359) fixes an issue

But, this means (to me) that the test should fail without the fix in #359, it does not, since that PR is open. So the test does not test what it claims to test.

In general, you should include a test with a PR, which #359 does not. So I would ask that you provide one PR with the main and test changes in the same PR, not split into two.

TY!

**Comment by StefanOltmann** [comment]
> But, this means (to me) that the test should fail without the fix in #359, it does not, since that PR is open. So the test does not test what it claims to test.

Please read carefully what I wrote here:

> Drop this [testfile.jpg](https://github.com/apache/commons-imaging/assets/9299303/2826b4ee-473a-45fb-b9fc-2168b9ae3fca) into src/test/resources/data/images/jpg to make the test fail.

You won't accept a PR with a failing test, would you?
So I did not include the test file.

And if I fix the test and the code in one PR, you will say "nothing changed".

The test now does not change, because none of the test files present raises the issue.

**Comment by StefanOltmann** [comment]
@garydgregory As you requested is #367 a combined PR of bugfix and repaired unit test plus a file that shows the problem.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:195]: Please do not add whitespace or change the style, it makes a PR larger, noisier, and takes longer to review.

**Comment by garydgregory** [comment]
A PR should include main and test changes, not one PR for test changes, and a different PR for main changes.

**Comment by StefanOltmann** [comment]
> A PR should include main and test changes, not one PR for test changes, and a different PR for main changes.

Yes, but the bugfix and the test are unrelated.

One PR fixes a bug without a test. The other PR completes a test that’s right now disabled, because it wasn’t working.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [reply src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:195]: Okay, won’t do that in future PRs.

**Comment by garydgregory** [comment]
Hi @kinow 
Any thoughts?

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:71]: Reading the code on my mobile, but this was the only part that called my attention, everything else looks good (can look with the IDE later today/this week).

IMO we should see if we can move these methods to test class (or internal package), and these should not be exposed/shipped as public API for Imaging.

But the update looks fine. Just need some moments with ide and coverage tool to confirm.

Thanks!

Comment (StefanOltmann) [reply src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:72]: @kinow, appreciate your review!

I understand the importance of being cautious when introducing a new API. Given the existence of a closely related method above (primarily utilized by tests), I believed its inclusion would be acceptable.

I've implemented the requested change, and it's now package private.

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:74]: Oh, a typo. Will fix in another PR.

File: src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java:19]: Will delete this class and simply use `Arrays.equals`. Can't see a reason to have the methods used for testing (and in one `if` statement in the code) while the other parsers use `Arrays.equals`. Please, do not update the PR with this, I have a working copy now.

Comment (kinow) [reply src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java:19]: (There is some code in `BinaryFunctions` related to this; anyway, for a future PR :+1: )

**Review by kinow — COMMENTED** [review]
Synced `master` (I was some hundreds of commits behind), and started reviewing but have to do some personal errands now. Will try to finish this one next week, and start looking at the 1.0 release in JIRA, to triage issues (and also see if there are other PR's). 

@tcurdt thanks a lot for the test! It's looking great! I will try to understand the changes you did, but you added comments which is really useful :slightly_smiling_face: 

I have a working copy with some modifications, so let's try not to touch it just for a few days :+1: 

Cheers

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (StefanOltmann) [reply src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:74]: I saw that too and decided it's out of scope. ;)

File: src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java
Comment (StefanOltmann) [reply src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java:19]: Sorry, but I can't follow you here.

The main issue is that the field `TiffOutputField.bytes` is private because it's an array that could be changed from outside the class. In Java we don't have immutable byte arrays unfortunately.

So the original `bytesEqual()` was created to compare an arbitrary byte array with the contents of the TiffOutputField. This is used in production code where the MakerNote protection is effective.

What we really need to test is a way to tell if two TiffOutputFields have the same content. There isn't a suitable other way to do so.

As you did not want to have additional public API this new class uses the package private way to access it from the test.

Another option would be to move the test into the same package.

But as I did not write the original test and did not want to change too much, I did not move it. Moving it would be a better option IMHO.

**Comment by StefanOltmann** [comment]
> I have a working copy with some modifications, so let's try not to touch it just for a few days 👍

If you test it locally don't forget to drop in the test file as described in my first post.
This will show you the problem my PR #359 fixes.

File: src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java
Comment (kinow) [reply src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java:19]: >There isn't a suitable other way to do so.

@StefanOltmann I think you are worrying right now without need. There are multiple ways to do it. We can, for instance, do as in other parts of Imaging and expose `getBytes()` returning a copy of the private array. This is done in other formats/parsers. I found only one place where one of those comparison methods was being used in the code.

>Another option would be to move the test into the same package.

:+1: agreed, there are multiple options, but nothing to worry here. Discussing about that shouldn't be necessary here. It's old code from Sanselan, that needs some attention before we release 1.0 final version.

There is nothing for you to worry, as there is nothing wrong in your code. No one is saying you changed too much, on the contrary, I told you you do not have to worry :slightly_smiling_face: I'm changing that because the code is old and instead of adding a new public method, or making it protected and adding an extra class, we may find other ways to solve it -- which is not pertinent to this.

If we digress too much on discussions like this, the pull request will quickly lose momentum, the committer working (myself in this case) will have to work on other things, and your pull request may be left with no updates for a while again.

So, please, do not worry about changing the code or justifying your changes, unless we request feedback for doing so :slightly_smiling_face: Your code is good, the API is old, the test looks great, just need to confirm the comments about offsets, and some changes you did in the test (which we do not need to discuss now, leave it be for a while until myself or another reviewer has had time to read it with calm)

:+1:

**Comment by kinow** [comment]
>If you test it locally don't forget to drop in the test file as described in my first post.

Noted :+1: You did a great job on the comments (in the code and in the pull request). We do not always get pull requests like this :slightly_smiling_face: Apologies for the delay, but wait just a bit more and this should be fixed.

For future PR's; it's common to have partial reviews. In some cases a reviewer might leave a note-to-self or to other committers (and maybe bring it to discuss over Slack or mailing list) in one of the comments, or may leave a partial review about something that, once the rest of the code is reviewed, does not need to be changed.

But once the review is done, you can be sure that someone will tag you, like "@StefanOltmann, I've reviewed it and it looks great! LGTM!", or "@StefanOltmann, can you look at my comment about the change you did in the code XPTO.java, please? I did not understand why you modified this or that".

Thanks again!

File: src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java
Comment (StefanOltmann) [reply src/test/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputFieldTestUtil.java:19]: Ok, great! :) 👍

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (garydgregory) [reply src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:72]: The method is still not package-private.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:142]: Don't change whitespace, please. It clutters up the PR with noise and makes it harder to review.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:177]: Don't clutter a PR changes in style, please, it clutters the PR with useless changes and takes longer to review.

**Review by garydgregory — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [reply src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:142]: It's a really small PR and this empty line makes the code easier to read.

File: src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java
Comment (StefanOltmann) [reply src/test/java/org/apache/commons/imaging/formats/jpeg/exif/ExifRewriterRoundtripTest.java:177]: It actually improves the code style. I see you doing a lot of code style changes in the commit history. So you should welcome a change like that.

File: src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java
Comment (StefanOltmann) [reply src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffOutputField.java:72]: @garydgregory It is. You are looking at the wrong commit.

These are the current lines:
```
    /* Only used in tests. */
    protected boolean bytesEqual(final TiffOutputField other) {
        return Arrays.equals(bytes, other.bytes);
    }
```

**Comment by StefanOltmann** [comment]
@garydgregory I won't do the changes you requested.

I'm waiting on a review from @kinow.

**Comment by StefanOltmann** [comment]
@garydgregory Let me quote myself from https://github.com/apache/commons-imaging/pull/367#issuecomment-1967723718:

> I think by now I did everything that I can do to contribute back something and you can either take my donation or leave it. That will be up to you.

I won't do any of the changes you requested.

**Comment by kinow** [comment]
- Replaced test code from `TiffOutputField` by a public getter/setter for the bytes data (the getter is a copy, like in other formats)
- Updated tests to use `assertArrayEquals` from JUnit instead of custom test code
- Updated a few lines to match the formatting of other classes
- Added link to the issue/PR where the test was added (for traceability)
- Added Javadocs to public methods changed

I added a second commit so we can still revert/remove things if needed, but if there are no objections I think this should be ready to be merged as soon as CI is happy (will take a look later, gotta go :wave: :runner: )

p.s. will hold merging this until the current ongoing release is out, I think, to avoid causing any issues with the release progress.

**Comment by kinow** [comment]
Build errors are on macos (works on macos + java 8, fails for java 11).

>Error:  Java heap space

I will spend some minutes checking if there's any Maven setting that we can use to fix the build, but I think that shouldn't be a blocker.

I still want to test this part from @StefanOltmann 's comment before merging

>Drop this [testfile.jpg](https://github.com/apache/commons-imaging/assets/9299303/2826b4ee-473a-45fb-b9fc-2168b9ae3fca) into src/test/resources/data/images/jpg to make the test fail.

Then I believe a good course of action here would be to merge this to get the test passing (thanks Stefan!), and continue working on #359. If I recall correctly we may not be able to include the test file, so I will check whether I can find another one, find another way to test that, or check with legal/others other options (no need to worry about this for now).

**Review by kinow — APPROVED** [review]
I ran the tests and all passed, then added the test file from the top comment to the `resources/data/images/tiff/15/fail.jpg`, re-ran the tests, and got one failure, as @StefanOltmann said it would happen in his top comment.

```bash
Expected :67
Actual   :0
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertArrayEquals.failArraysNotEqual(AssertArrayEquals.java:440)
	at org.junit.jupiter.api.AssertArrayEquals.assertArrayEquals(AssertArrayEquals.java:209)
	at org.junit.jupiter.api.AssertArrayEquals.assertArrayEquals(AssertArrayEquals.java:63)
	at org.junit.jupiter.api.Assertions.assertArrayEquals(Assertions.java:1238)
	at org.apache.commons.imaging.formats.jpeg.exif.ExifRewriterRoundtripTest.assertTiffEquals(ExifRewriterRoundtripTest.java:131)
	at org.apache.commons.imaging.formats.jpeg.exif.ExifRewriterRoundtripTest.updateExifMetadataLossless_copyWithoutChanges_TiffOutputSetsAreIdentical(ExifRewriterRoundtripTest.java:245)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

So +1 to merge this so that the existing `@Disabled` test can be re-enabled, and so we can continue working on the #359 PR.

@garydgregory I won't merge it now as you had a release vote recently, so I don't want to cause any issues with the release process. So feel free to review/merge later at any time :+1: 

I left a second commit here in case we need to revert anything, but it can be squashed (just taking care to keep Stefan's original commit).

Thanks @StefanOltmann , and sorry the delay, I took some days off before Easter to have a very long holiday, catching up with issues now.

**Comment by kinow** [comment]
And tests passed after re-running the failed jobs on GitHub Actions :man_shrugging:

**Comment by garydgregory** [comment]
@kinow 
Thank you for your work here. I'll merge this once 1.0.0-alpha4 is out (in 3 days).

**Comment by kinow** [comment]
Rebased, added `changes.xml` entry, and squashed commits. The `changes.xml` has the due-to field set to Stefan and to Gary Lucas (as I believe Stefan used Gary's suggestion from JIRA :+1:).

Waiting for GH Actions now to complete before merging.

**Comment by StefanOltmann** [comment]
@kinow Thanks for merging it! :)

> The `changes.xml` has the due-to field set to Stefan and to Gary Lucas (as I believe Stefan used Gary's suggestion from JIRA 👍).

That's true for #359. This PR fixed & completed work started by Charles Hope in #275. But yes, this test was built in the first place to showcase the bug that Gary found the bugfix for - so it's fine to give him credit. :)

**Comment by StefanOltmann** [comment]
@kinow Thanks for merging it. :)

> The changes.xml has the due-to field set to Stefan and to Gary Lucas (as I believe Stefan used Gary's suggestion from JIRA 👍).

This PR completes the work done in PR #275, started by Charles Hope.
But since Gary Lucas found the bugfix and the test was done to showcase the bug it fixes, it's fine to give him credit. :)

**Comment by StefanOltmann** [comment]
I don't know eactly the policy is on the `changes.xml`, but I just noticed that there seems to be no entry for the initial `ExifRewriterRoundtripTest`.

So to give everyone involved proper credit maybe "Charles Hope" should be mentioned.

**Comment by kinow** [comment]
> I don't know eactly the policy is on the `changes.xml`, but I just noticed that there seems to be no entry for the initial `ExifRewriterRoundtripTest`.
> 
> So to give everyone involved proper credit maybe "Charles Hope" should be mentioned.

Sounds good to me! I will do some searching in the git history to confirm where it should be added. Thanks @StefanOltmann !

**Pull request merged**
